modules/tensor_mechanics/src/kernels/AsymptoticExpansionHomogenizationKernel.C

```diff
@@ -0,0 +1,97 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AsymptoticExpansionHomogenizationKernel.h"
+
+registerMooseObject("TensorMechanicsApp", AsymptoticExpansionHomogenizationKernel);
+
+template <>
+InputParameters
+validParams<AsymptoticExpansionHomogenizationKernel>()
+{
+  InputParameters params = validParams<Kernel>();
+  params.addClassDescription("Kernel for asymptotic expansion homogenization for elasticity");
+  params.addRequiredRangeCheckedParam<unsigned int>("component",
+                                                    "component >= 0 & component < 3",
+                                                    "An integer corresponding to the direction "
+                                                    "the variable this kernel acts in. (0 for x, "
+                                                    "1 for y, 2 for z)");
+  MooseEnum column("xx yy zz yz xz xy");
+  params.addRequiredParam<MooseEnum>("column",
+                                     column,
+                                     "The column of the material matrix this kernel acts in. "
+                                     "(xx, yy, zz, yz, xz, or xy)");
+
+  params.addParam<std::string>("base_name",
+                               "Optional parameter that allows the user to define "
+                               "multiple mechanics material systems on the same "
+                               "block, i.e. for multiple phases");
+
+  return params;
+}
+
+AsymptoticExpansionHomogenizationKernel::AsymptoticExpansionHomogenizationKernel(
+    const InputParameters & parameters)
+  : Kernel(parameters),
+
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _elasticity_tensor(getMaterialPropertyByName<RankFourTensor>(_base_name + "elasticity_tensor")),
+    _component(getParam<unsigned int>("component")),
+    _column(getParam<MooseEnum>("column"))
+{
+}
+
+Real
+AsymptoticExpansionHomogenizationKernel::computeQpResidual()
+{
+  unsigned k = 0, l = 0;
+
+  if (_column == 0)
```

Comment:
This would be more compact with a lookup from an array

```
std::array<unsigned int, 6> _k_index; 
...
_k_index({0, 1, 2, 1, 0, 0}),
...
k = _k_index[_component];
```

and so on